packages/patternfly-4/react-core/src/components/Form/FormGroup.js

```diff
@@ -0,0 +1,121 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { FormContext } from '../Form/FormContext';
+import { css, getModifier } from '@patternfly/react-styles';
+import styles from '@patternfly/patternfly-next/components/Form/styles.css';
+
+const propTypes = {
+  /** FormGroup field. */
+  children: PropTypes.node,
+  /** Custom class name. */
+  className: PropTypes.string,
+  /** Label text before the field. */
+  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
+  /** Sets the FormGroup required. */
+  required: PropTypes.bool,
+  /** Sets the FormGroup invalid. */
+  invalid: PropTypes.bool,
+  /** Helper text after the field. It can be a simple text, a function or an object. */
+  helpertext: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
+  /** Helper text after the field when the field is invalid. It can be a simple text, a function or an object. */
+  helpertextInvalid: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
+  /** ID of the included field. It has to be the same for proper working. */
+  fieldId: PropTypes.string
+};
+
+const defaultProps = {
+  children: null,
+  className: '',
+  label: undefined,
+  required: false,
+  invalid: false,
+  helpertext: undefined,
+  helpertextInvalid: undefined,
+  fieldId: undefined
```

Comment:
the id should be required like it is in core.

Reply:
Well, not quite all the time. The id should be required if label is text, for other use cases it's not really required. However you are correct that it should be required if user wants to use text as label, so function should be used here.
```JSX
const propTypes = {
  ...
  id: (props) => {
    if (props.label && typeof props.label === 'string' && !props.id) {
      return new Error('FormGroup requires ID if label is string.');
    }
  }
}
```

Reply:
I added the function!

Reply:
This question might be more related to [previous comments](https://github.com/patternfly/patternfly-react/pull/667#pullrequestreview-159609445) that are now outdated about allowing the label to accept values other than string... But is it possible to catch and warn the consumer about possible accessibility issues? For example, if something other than a string is passed for the label, can we warn the consumer that they should check that text is included somewhere inside the `<label>` (because without providing an accessible form label that the form will fail accessibility checks)?

> The id should be required if label is text, for other use cases it's not really required. 

If the consumer does provide a label that is not text, then the input should have an `aria-label` attribute that includes the text label for the form field. Can we catch this case and warn the consumer that one of the following must be true:

1. `<label>` includes a text string and the `<input>` includes an `id`
    _or_
2. `<input>` includes an `aria-label`

Reply:
Good idea! 

FormGroup throws an error now if:

1. **label is not a string**

- Children is an **array** and no child has an` aria-label` (at least one of the children has to have an `aria-label`)
- Child does not have an `aria-label`

2. **label is a string**
- Children is an **array** and no child has an` id` (at least one of the children has to have an `id`)
- Child does not have an `id`

Reply:
Great Updates! 
They only concern I have is  that you are only warning that an aria-label/id is needed if none of the children has one.  In order for the component to be accessible, all the visible children should have an aria label if the Label is not a string.  And if the label is a string, all the visible children should have an id.

Reply:
To make sure I understand the FormGroup, it sounds like it can have one label and more than one input element?

Each input element should have a unique text string associated with it. And a text string can be associated with an input element using the `for` attribute on a `<label>` or the `aria-label` attribute on the input itself. 

I agree with @tlabaj's comment above about catching potential a11y issues and warning the user. I just wanted to expand on this statement: 

> And if the label is a string, all the visible children should have an id.

A single label can only be associated with a single input element using the `for`/`id` relationship. Therefore, I think the correct logic would be:

- If the label is a string, and only one visible child in the array, then that child must have an id
- If the label is a string, and more than one visible child in the array, then all children must have aria-label with the exception of the first child. The first child does not need `aria-label` if the label `for` attribute references the `id` value of that child. 
  - _It could be argued that we could allow any one child to be referenced by the label and not limit it to the first child. It would just be weird from the user experience perspective._

Reply:
It has not to be only input elements. For example, I can imagine to put there an input with some image. And in cases like that you don’t need id or aria-label on that image:

```
<FormGroup label='How would you name this cat?' fieldId='cat-id' >
     <img src=‘imageofawesomecat.png' />
     <TextInput id='cat-id’ />
</ FormGroup>
```

In ideal approach FormGroup is just styling wrapper around some input field (select, textInput, etc.) with support of the helperText and label. Nothing more.

And if not, I would let consumers to add more labels, if they want more inputs. You can use the component like that (however I think it does not make sense to use it with more inputs):

```
<FormGroup label='How would you name this cat?' fieldId='cat-id' >
     <img src=‘imageofawesomecat.png' />
     <TextInput id='cat-id’ />
     <Label for='cat2'>And any last name?</ Label>
     <TextInput id='cat-last-name' />
</ FormGroup>
```
There will be probably some (and big) problems with styling, and consumers should manage it themselves. But it's up to them. I think it is more about to give consumers more freedom how to use the component if they want to. But I'm not a designer or architect, so I don't know how strict we want to be or how much freedom we want to allow. And of course, everything of this should be mentioned in a consumer documentation.

Reply:
Thanks for the input (no pun intended), @rvsia and @Hyperkid123! The best suggestion I have at this time is to move forward with this PR, and then later revisit this in the context of a demo or sample app, and just test what the consumer experience is for common use cases. I think that will help us understand if there are any gaps in how we inform consumers, and also how to best address them.